core/include/vecmem/memory/unique_ptr.hpp

```diff
@@ -0,0 +1,462 @@
+#include <type_traits>
+
+#include "vecmem/memory/memory_resource.hpp"
+
+namespace vecmem {
+namespace {
+/**
+ * @brief A deleter class for non-trivial objects.
+ *
+ * This class provides all the necessary functionality to allow unique pointers
+ * to deallocate non-trivial objects, as well as arrays of them.
+ *
+ * Object pointers not only deallocate their memory, but they also cleanly
+ * delete the object by callings its deconstructor.
+ *
+ * The design of this class is somewhat unconventional because it is stateful,
+ * which most deleters are not. However, this is pretty much necessary we
+ * cannot rely on the packed storage of allocation parameters that you see in
+ * standard deleters. In particular, the pointer argument is not strictly
+ * necessary, but it is included to make the class easier to debug.
+ *
+ * @tparam T The type to deallocate.
+ *
+ * @note This deleter class supports non-array types and array types of unknown
+ * bound, but it does not support arrays of known bound.
+ *
+ * @warning This type should never be used with host-inaccessible memory! We
+ * cannot check this at compile-time or at run-time, so it is the
+ * responsibility of the user to ensure that this requirement is respected.
+ */
+template <typename T,
+          std::enable_if_t<!std::is_array_v<T> ||
+                               (std::is_array_v<T> && std::extent_v<T> == 0),
+                           bool> = true>
+struct unique_obj_deleter {
+public:
+    using pointer_t =
+        std::conditional_t<std::is_array_v<T>, std::decay_t<T>, T*>;
+    using storage_t = std::remove_pointer_t<pointer_t>;
+
+    /**
+     * @brief Default-construct a new unique object deleter.
+     *
+     * Having a default constructor for this class is not ideal, as having an
+     * empty deleter here doesn't really make any sense. However, if the
+     * deleter class is not default-constructible, then neither is the
+     * std::unique_ptr class using it. For that reason, we want it here.
+     */
+    unique_obj_deleter(void) = default;
+
+    /**
+     * @brief Construct a new unique object deleter.
+     *
+     * This is the preferred constructor for this type. It will store a memory
+     * resource, a pointer, a size, and possible an alignment and a number of
+     * elements.
+     *
+     * @param mr The memory resource to use for deallocation.
+     * @param p The pointer which we will want to deallocate.
+     * @param s The size of the allocation.
+     * @param a The alignment of the allocation, with 0 representing no
+     * alignment.
+     * @param n The number of elements in the allocation.
+     */
+    unique_obj_deleter(memory_resource& mr, pointer_t p, std::size_t s,
+                       std::size_t a = 0, std::size_t n = 1)
```

Comment:
Shouldn't we just take the default alignment from [std::pmr::memory_resource](https://en.cppreference.com/w/cpp/memory/memory_resource)? I.e.:
  - https://en.cppreference.com/w/cpp/memory/memory_resource/allocate
  - https://en.cppreference.com/w/cpp/memory/memory_resource/deallocate

```c++
alignof(std::max_align_t)
```

Then there would be no reason for the `if (m_align > 0)` check later on.

Reply:
I think I would prefer to keep it this way: once we do the allocation, there is no way to re-obtain the alignment later, and it is a required argument to the deallocate function. Right now, there are no cases where we don't use the default alignment, but I think it would be a solid idea to keep the possibility open to support this in the future. In general, I think we've been a little (too) lax in supporting alignment throughout vecmem as a whole, and I am quite worried that this will come bite us in the behind at some point.

On the other hand the YAGNI principle would dictate not having this, so I am happy to go either way.

Reply:
I think you misunderstood me... I don't want to get rid of the `m_align` variable. I was just proposing to replace `0` with `alignof(std::max_align_t)` as its default value in the constructor.

Then the deallocation could simply write:

```c++
m_mr->deallocate(p, m_size, m_align);
```

Instead of the current:

```c++
        if (m_align > 0) {
            m_mr->deallocate(p, m_size, m_align);
        } else {
            m_mr->deallocate(p, m_size);
        }
```

Reply:
It would be possible, but I must admit I don't really see the point. The only thing I see this changing is that you start making some assumptions on the default value of `std::pmr::memory_resource::allocate`'s default alignment argument - if those change for whatever reason (unlikely, sure, but not impossible) then suddenly the deleter is no longer correct. The current implementation treats the memory resource as the single source of truth, rather than making these assumptions. What would you say is the advantage of having `std::max_align_t` be the default?

Reply:
:thinking: I assumed that if cppreference spelled it out, the default argument value might be part of the standard. But the standard indeed does not seem to define what the default value should be...

http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/n4659.pdf

(Section 23.12.2.)

So fine, let's use 0 as the default/undefined value then...